### DISCUSSION TITLE
How to make a map using echarts in nicegui

### DISCUSSION THREAD

**Discussion by CrystalWindSnake** [zauberzeug/nicegui#2407] [Q&A] [answered]
### Question

echarts requires the following steps to create a map

1. Call the `registerMap` function of the global echarts object to register the map data.
2. Use the registered map name in the chart component.

how do I make sure that the map registration (point 1 above) is performed before the chart component is instantiated?

I found a rather hacky implementation. 
```python
from nicegui import app, ui

ui.add_body_html(
    r"""
<script>

window.addEventListener('DOMContentLoaded', () => {

    fetch("https://geo.datav.aliyun.com/areas_v3/bound/100000_full.json")
    .then((response) => response.json())
    .then((data) => {
        echarts.registerMap('china',data)
    });  

});
</script>
"""
)

opts = {
    "geo": {
        "map": "china",
        "roam": True,
    },
    "tooltip": {},
    "legend": {},
    "series": [],
}

echart = ui.echart({})

def setup_chart():
    echart._props["options"] = opts
    echart.update()

ui.timer(2, setup_chart, once=True)

ui.run()
```

Is there a more reasonable way?

**Comment by falkoschindler** [comment, marked as answer]
Hi @CrystalWindSnake,

that's a tricky problem. Here's a possible approach based on a custom "map loaded" event:

```py
@ui.page('/')
def page():
    ui.add_body_html(r'''
        <script>
            window.addEventListener('DOMContentLoaded', () => {
                fetch("https://geo.datav.aliyun.com/areas_v3/bound/100000_full.json")
                    .then((response) => response.json())
                    .then((data) => {
                        echarts.registerMap('china', data);
                        emitEvent('map_loaded');
                    });
            });
        </script>
    ''')
    echart = ui.echart({'series': []})
    ui.on('map_loaded', lambda: (
        echart.options.update(geo={'map': 'china', 'roam': True}),
        echart.update(),
    ))
```

Note that it requires a page function to work.

**Reply by CrystalWindSnake** [reply]
Thanks.
it only works in the page function, is that because of `ui.on`? I didn't find documentation for `ui.on` on the nicegui website.

**Reply by falkoschindler** [reply]
Without a page function all browser tabs share the same state. The first tab loads the JSON, emits the event and adds the map "china" to the ECharts options. When the second tab connects (or the first tab reloads), the EChart options already contain the map "china", but the client hasn't loaded the JSON. This causes the JavaScript error. With a page function every tab has its own ECharts options and the problem doesn't occur.

**Reply by CrystalWindSnake** [reply]
I get it. Thanks for explaining.
